source/simulator/solver_schemes.cc

```diff
@@ -431,12 +431,19 @@ namespace aspect
         compute_current_constraints ();
       }
 
-    // the Stokes matrix depends on the viscosity. if the viscosity
-    // depends on other solution variables, then after we need to
-    // update the Stokes matrix in every time step and so need to set
-    // the following flag. if we change the Stokes matrix we also
-    // need to update the Stokes preconditioner.
-    rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+    // If the Stokes matrix depends on the solution, or we have active
+    // velocity boundary conditions, we need to re-assemble the system matrix
+    // (and preconditioner) every time. If we have active boundary conditions,
+    // they could a) depend on the solution, or b) be inhomogeneous. In both
+    // cases, just assembling the RHS will be incorrect.  If no active
+    // boundaries exist, we only have no-slip or free slip conditions, so we
+    // don't need to force assembly of the matrix.
+    if (stokes_matrix_depends_on_solution()
+        ||
+        (nonlinear_iteration == 0 && boundary_velocity_manager.get_active_boundary_velocity_conditions().size() > 0))
+      rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+    else if (parameters.enable_prescribed_dilation)
+      rebuild_stokes_matrix = true;
```

Comment:
can you leave a comment about this? I guess this modifies the RHS, so we need to rebuild?

Reply:
also, why do we not need to do rebuild_stokes_preconditioner?

Reply:
I am not entirely sure what to comment on it, since I do not really understand what the prescribed dilation is trying to do, and why it needs it, but for example the elasticity doesn't. To be honest, the main reason I added this line is that it is the minimal change needed to fix the dilation test. So my main question to you is whether that test is good enough, I might I have broken something with the dilation by doing this?

Reply:
Does https://github.com/geodynamics/aspect/blob/fcf14966ccf61e3fac625bc0ccfbbe832953db9a/include/aspect/material_model/interface.h#L1061-L1073 help? It modifies the RHS (potentially in every iteration).
Leaving a comment that prescribed_dilation modifies the RHS would be helpful.

Reply:
Ok, I will add a comment to it. So my guess is that only rebuild_stokes_matrix is needed because we only update the rhs, and the lhs doesn't change.

Reply:
I don't understand "the Stokes matrix (which is on the rhs in the Newton solver)".

Reply:
with the Stokes matrix I refer to `A` in `J=Ax-b`, and `J` is the `Newton stokes matrix` in de code.